packages/devextreme/js/__internal/data_controller.ts

```diff
@@ -0,0 +1,265 @@
+import { extend } from '../core/utils/extend';
+import ArrayStore from '../data/array_store';
+import { DataSource } from '../data/data_source/data_source';
+import { normalizeDataSourceOptions } from '../data/data_source/utils';
+
+interface DataSourceType {
+  _userData: unknown;
+  _pageSize: number;
+  dispose: () => void;
+  load: (loadOptions?: unknown) => Promise<unknown>;
+  loadSingle: (propName: string, propValue: unknown) => Promise<unknown>;
+  loadOptions: () => unknown;
+  cancel: (operationId: number) => void;
+  cancelAll: () => void;
+  filter: (filter?: unknown) => unknown;
+  group: (group?: unknown) => unknown;
+  items: () => unknown[];
+  key: () => string;
+  keyOf: (item: unknown) => string;
+  paginate: (paginate?: unknown) => unknown;
+  pageIndex: <TIdx = number | undefined>(pageIndex: TIdx) => TIdx extends number
+    ? undefined
+    : number;
+  reload: () => Promise<unknown>;
+  select: (...args: unknown[]) => unknown;
+  searchExpr: (searchExpr?: unknown) => unknown;
+  searchOperation: (searchOperation?: string) => unknown;
+  searchValue: (searchValue?: unknown) => unknown;
+  store: () => DataSourceType;
+  totalCount: () => number;
+  isLastPage: () => boolean;
+  isLoaded: () => boolean;
+  isLoading: () => boolean;
+  on: (eventName: string, eventHandler: () => void) => DataSourceType;
+  off: (eventName: string, eventHandler: () => void) => DataSourceType;
+  _addSearchFilter: (storeLoadOptions: unknown) => void;
+  _applyMapFunction: <TData = unknown>(data: TData) => TData;
+}
+
+class DataController {
+  private _isSharedDataSource = false;
+
+  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+  // @ts-expect-error
+  private _dataSource: DataSourceType;
+
+  constructor(dataSourceOptions: DataSourceType) {
+    this._initDataSource(dataSourceOptions);
+  }
+
+  static initFromArray(items: unknown[], key: string): DataController {
+    const dataSource = new DataSource({
+      store: new ArrayStore({
+        key,
+        data: items,
+      }),
+      pageSize: 0,
+    });
+    return new DataController(dataSource as unknown as DataSourceType);
+  }
+
+  _initDataSource(dataSourceOptions: DataSourceType): void {
```

Comment:
I think `update` is better here because we start this function with a restoring of a previous dataSource
```suggestion
  _updateDataSource(dataSourceOptions: DataSourceType): void {
```